### PR TITLE
Modernize the codebase - remove leftover NULLs

### PR DESCRIPTION
**Pull request by JanVogelsang**
There were some occurrences of `NULL` keywords that were not picked up by `clang-tidy` and are now added in this PR.

### CODE REVIEW

**Comment by heplesser** [comment]
@JanVogelsang I restarted one failed test, it seemed to be a technical issue on the runner. I would much prefer to put this PR in draft mode until #2498 is merge and base this PR on code with #2498 integrated. Otherwise it becomes more difficult to review and there is a small risk that merging this one would reintroduce && and II if diff should get confused.

File: libnestutil/config.h.in [hunk 1]
```diff
@@ -129,7 +129,7 @@
 /* "Define if NAN is available" */
 #cmakedefine HAVE_NAN 1
 
-/* "Define if std::nan(NULL) is available" */
+/* "Define if std::nan(nullptr) is available" */
```
Comment (heplesser) [inline libnestutil/config.h.in:132]:
```suggestion
/* "Define if std::nan(const char*) is available" */
```
(see https://en.cppreference.com/w/cpp/numeric/math/nan) or just
```suggestion
/* "Define if std::nan() is available" */
```

As an aside: We now require C++11, and this function is guaranteed to be in C++11. Therefore, do we need this test? And which other tests and pertaining conditionals could we remove? That should be a follow-up issue.

File: libnestutil/lockptr.h [hunk 2]
```diff
@@ -114,7 +114,7 @@ class lockPTR
     ~PointerObject()
     {
       assert( not locked );
-      if ( ( pointee != NULL ) && deletable && ( not locked ) )
+      if ( pointee && deletable && ( not locked ) )
```
Comment (heplesser) [inline libnestutil/lockptr.h:117]:
```suggestion
      if ( pointee and deletable and not locked )
```

File: libnestutil/lockptr.h [hunk 3]
```diff
@@ -234,41 +234,41 @@ class lockPTR
...
     return *( obj->get() );
   }
 
 
   bool
-  operator not() const //!< returns true if and only if obj->pointee == NULL
+  operator not() const //!< returns true if and only if not obj->pointee
   {
-    // assert(obj != NULL);
+    // assert(obj);
```
Comment (heplesser) [inline libnestutil/lockptr.h:269]: Either delete or uncomment.

File: sli/sliactions.cc [hunk 4]
```diff
@@ -115,7 +115,7 @@ TrietypeFunction::execute( SLIInterpreter* i ) const
 void
 CallbacktypeFunction::execute( SLIInterpreter* i ) const
 {
-  //    assert(i->ct.datum() != NULL); // we wouldn't be here otherwise
+  //    assert(i->ct.datum()); // we wouldn't be here otherwise
```
Comment (heplesser) [inline sli/sliactions.cc:118]:
```suggestion
```

File: sli/sliarray.cc [hunk 5]
```diff
@@ -106,7 +106,7 @@ SLIArrayModule::RangeFunction::execute( SLIInterpreter* i ) const
   {
     IntegerDatum* n1d = dynamic_cast< IntegerDatum* >( ad->get( 0 ).datum() );
     IntegerDatum* n2d = dynamic_cast< IntegerDatum* >( ad->get( 1 ).datum() );
-    if ( ( n1d ) && ( n2d ) )
+    if ( n1d && n2d )
```
Comment (heplesser) [inline sli/sliarray.cc:109]:
```suggestion
    if ( n1d and n2d )
```
Applies also below.

File: sli/sliarray.cc [hunk 6]
```diff
@@ -1613,26 +1613,26 @@ void
 SLIArrayModule::IMapThreadFunction::execute( SLIInterpreter* i ) const
 {
   ProcedureDatum* procd = static_cast< ProcedureDatum* >( i->EStack.pick( 1 ).datum() );
-  //    assert(procd != NULL);
+  //    assert(procd);
```
Comment (heplesser) [inline sli/sliarray.cc:1616]: Remove all commented-out code.

File: sli/tarrayobj.cc [hunk 7]
```diff
@@ -628,9 +628,9 @@ TokenArrayObj::operator==( const TokenArrayObj& a ) const
   }
 
   // experimentally replaced by line below 090120, Diesmann
```
Comment (heplesser) [inline sli/tarrayobj.cc:630]: Remove

**Review by heplesser — CHANGES_REQUESTED** [review]
Looks good in general, but see comments. All commented-out code should be removed (in this PR only lines you touch), we have version control to keep an archive of what was. I also spotted some &&.

File: libnestutil/lockptr.h (hunk 2 above)
Comment (JanVogelsang) [reply libnestutil/lockptr.h]: Resolved after merging with master

File: sli/sliarray.cc (hunk 5 above)
Comment (JanVogelsang) [reply sli/sliarray.cc]: Resolved after merge with master

File: libnestutil/config.h.in (hunk 1 above)
Comment (JanVogelsang) [reply libnestutil/config.h.in:132]: I created an issue #2508, as I agree that we should simply remove this check now, as well as some others

**Review by jougs — COMMENTED** [review]
Thanks! Other than my few comments, I'm fine with this.

One general question now that I come to think of it: Did any of your recent cleanup PRs touch code in the `thirdparty` directory? Our usual policy is to leave those as is in order to match the upstream code as good as possible. They are also excluded from our static code checks in the CI.

File: doc/htmldoc/static/js/11.1.0_highlight.min.js [hunk 8]
```diff
@@ -419,7 +419,7 @@ className:"title",begin:u(t)+e.IDENT_RE,relevance:0
 },c=u(t)+e.IDENT_RE+"\\s*\\(",d={
 keyword:["asm","auto","break","case","const","continue","default","do","else","enum","extern","for","fortran","goto","if","inline","register","restrict","return","sizeof","static","struct","switch","typedef","union","volatile","while","_Alignas","_Alignof","_Atomic","_Generic","_Noreturn","_Static_assert","_Thread_local","alignas","alignof","noreturn","static_assert","thread_local","_Pragma"],
 type:["float","double","signed","unsigned","int","short","long","char","void","_Bool","_Complex","_Imaginary","_Decimal32","_Decimal64","_Decimal128","complex","bool","imaginary"],
-literal:"true false NULL",
+literal:"true false nullptr",
```
[the remark this answers is not in the document]
Comment (JanVogelsang) [reply doc/htmldoc/static/js/11.1.0_highlight.min.js:422]: Thought about having both as well, then after second thought removed the NULL entirely, as there should not be any NULLs in the codebase anyway. Still, it wouldn't hurt addin the NULL of course.

[the remark this answers is not in the document]
Comment (jougs) [reply doc/htmldoc/static/js/11.1.0_highlight.min.js:422]: I guess it comes from here: https://github.com/highlightjs/highlight.js

@JanVogelsang: please create a PR there. Pulling upstream into our repo could be a separate PR then.

**Review by heplesser — CHANGES_REQUESTED** [review]
Just a few details left ;).

File: doc/htmldoc/static/js/11.1.0_highlight.min.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jougs) [reply doc/htmldoc/static/js/11.1.0_highlight.min.js:422]: Also good. Thanks for looking into this.

File: sli/typechk.cc [hunk 9]
```diff
@@ -255,7 +252,7 @@ TypeTrie::insert_move( const TypeArray& a, Token& f )
   /* Error conditions:
      1. If pos->next!=NULL, the parameter list overlaps with
```
Comment (heplesser) [inline sli/typechk.cc:253]: @JanVogelsang Just one little NULL left ;)
```suggestion
     1. If pos->next != nullptr, the parameter list overlaps with
```

**Review by heplesser — APPROVED** [review]
@JanVogelsang Thanks, all fine from my side now!

**Review by jougs — APPROVED** [review]
LGTM!

**Pull request merged**